dev-docs/RFCs/v7.x/python-integration-rfc.md

```diff
@@ -0,0 +1,117 @@
+
+By providing a set of very simple bindings, focusing on overcoming the most basic integration issues and exposing a reasonable subset of the deck.gl API, we should be able to provide significantly increase adoption and value of deck.gl in these areas.
+
+Offering such an integration clearly makes sense, as long as we can do it with low effort and maintenance cost.
+
+
+### Leveraging the deck.gl JSON API
+
+Normally, exposing an API in a new language requires porting that API to the new language, which can be a significant undertaking, creating a large set of class and function "wrappers" that glue to the original code, with associated maintenance costs.
+
+However, deck.gl now supports a JSON API. The JSON API makes it possible for users to specify deck.gl layers and access a large subset of deck.gl's functionality without directly using the deck.gl JavaScript API, making it an ideal API entry point to expose in other languages.
+
+The idea is thus to create a minimal binding in Python that allows a python application to generate a JSON payload compatible with the deck.gl API and then start an inline frame running deck.gl rendering that JSON payload.
+
+
+### Focus Areas
+
+To minimize initial effort and long-term maintenance, we want to keep the scope very limited and focus on solving core problems such as:
+
+* Making the deck.gl Python module easy to install and use for Python users
+* Handle proper initial sizing and resizing the JavaScript/HTML window
+* Efficiently transfer large data sets between Python and JavaScript
+* Make it easy to build more ambitions Python class libraries on top of the basic module.
+
+
+### Constraints
+
+deck.gl should not grow to include a major Python port of its API (e.g. a Python class library that mirrors all deck.gl layers). That would represent too big of a distraction to the core deck.gl mission. We should leave such work to the community, and instead encourage, support and promote such efforts as they happen.
+
+Something we would likely want to support is that other deck.gl/Python integtation libraries build on our official integration (i.e. they use it as a python dependency). Alternatively, they could just use it as an inspiration/springboard to get started, or fork it to help them solve some of the initial integration issues.
+
+
+## Technical Challenges
+
+* Starting up a JavaScript iframe in Notebook
+
+* Controlling Size of iframe
+  * full screen rendering
+  * height of frame controllable by app?
+  * frame resizable by user?
+
+* One way communication
+  * Most Python/JS bindings generate HTML dynamically using Python API then create the iframe, after that Python has no more control.
+  * However deck.gl JSON api can accept new JSON payloads, they go through standard deck.gl layer diffing.
+  * Maintain a network connection so that Python can push new payloads?
+
+* Efficient data transfer
+  * Many Python JS integrations serialize data into the HTML/JS payload.
```

Comment:
Is it possible for the Python end to start a server that hosts/updates a static data file?

Reply:
Not within Jupyter notebook, I assume, unless we modify the jupyter notebook [server](https://jupyter-notebook.readthedocs.io/en/stable/public_server.html). But users will have to use our version of jupyter notebook to leverage such capability.

>   * Many Python JS integrations serialize data into the HTML/JS payload.
@ibgreen by serializing data into the payload you mean the data is embedded in the HTML file? We can pass in the data on-the-fly via a thin python wrapper like the way we pass props to React components. ([e.g.](https://github.com/rulematrix/rule-matrix-py/blob/master/rulematrix/visualization/data_model.py#L218))

Reply:
Looked at the example you linked, following the render function call into the other file. I am not 100% sure but to me it looks like it just stringifies the JSON data into the generated HTML.